File: libs/opensource/sphinx/plugins/org.eclipse.sphinx.emf/src/org/eclipse/sphinx/emf/ecore/proxymanagement/AbstractProxyResolverService.java
/**
 * <copyright>
 *
 * Copyright (c) 2014-2017 itemis and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/org/documents/epl-2.0/EPL-2.0.html
 *
 * Contributors:
 *     itemis - Initial API and implementation
 *     itemis - [475954] Proxies with fragment-based proxy URIs may get resolved across model boundaries
 *     itemis - [501899] Use base index instead of IncQuery patterns
 *
 * </copyright>
 */
package org.eclipse.sphinx.emf.ecore.proxymanagement;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import org.eclipse.emf.common.util.URI;
import org.eclipse.emf.ecore.EClass;
import org.eclipse.emf.ecore.EObject;
import org.eclipse.sphinx.emf.metamodel.IMetaModelDescriptor;
import org.eclipse.sphinx.emf.metamodel.services.AbstractMetaModelService;
import org.eclipse.sphinx.emf.resource.ExtendedResourceSet;

public abstract class AbstractProxyResolverService extends AbstractMetaModelService implements IProxyResolverService {

	private List<IProxyResolver> proxyResolvers = new ArrayList<IProxyResolver>();

	public AbstractProxyResolverService(Collection<IMetaModelDescriptor> mmDescriptors) {
		super(mmDescriptors);
		initProxyResolvers();
	}

	protected abstract void initProxyResolvers();

	protected EClass getTargetEClass(URI uri) {
		return null;
	}

	protected List<IProxyResolver> getProxyResolvers() {
		if (proxyResolvers == null) {
			proxyResolvers = new ArrayList<IProxyResolver>();
		}
		return proxyResolvers;
	}

	protected IProxyResolver getProxyResolver(EClass eType) {
		for (IProxyResolver resolver : getProxyResolvers()) {
			if (resolver.canResolve(eType)) {
				return resolver;
			}
		}
		return null;
	}

	@Override
	public EObject getEObject(EObject proxy, EObject contextObject, boolean loadOnDemand) {
		IProxyResolver proxyResolver = getProxyResolver(proxy.eClass());
		if (proxyResolver != null) {
			return proxyResolver.getEObject(proxy, contextObject, loadOnDemand);
		}
		return null;
	}

	@Override
	public EObject getEObject(URI uri, ExtendedResourceSet contextResourceSet, Object contextObject, boolean loadOnDemand) {
		EClass targetEClass = getTargetEClass(uri);
		if (targetEClass != null) {
			IProxyResolver proxyResolver = getProxyResolver(targetEClass);
			if (proxyResolver != null) {
				return proxyResolver.getEObject(uri, targetEClass, contextResourceSet, contextObject, loadOnDemand);
			}
		}
		return null;
	}
}

File: cpp/emf-cpp/emf-sphinx/src/ecore/proxymanagement/AbstractProxyResolverService.cpp
// AbstractProxyResolverService.cpp
#include "emf/sphinx/ecore/proxymanagement/AbstractProxyResolverService.h"
#include "emf/sphinx/metamodel/IMetaModelDescriptor.h"
#include "emf/common/EObject.h"
#include "emf/common/Resource.h"
#include "emf/common/Resource.h"
#include "emf/common/URI.h"
#include "emf/sphinx/resource/ExtendedResourceSet.h"

namespace emf::sphinx::ecore::proxymanagement {

AbstractProxyResolverService::AbstractProxyResolverService(
    const std::vector<emf::sphinx::metamodel::IMetaModelDescriptor*>& mmDescriptors)
    : mmDescriptors_(mmDescriptors) {}

bool AbstractProxyResolverService::isApplicableTo(emf::sphinx::metamodel::IMetaModelDescriptor* mm) const {
    if (!mm) return false;
    for (auto* d : mmDescriptors_) {
        if (d == mm) return true;
    }
    return false;
}

emf::common::EObject* AbstractProxyResolverService::getEObject(emf::common::EObject* /*proxy*/,
                                                                emf::common::EObject* /*context*/, bool /*loadOnDemand*/) {
    return nullptr;
}

emf::common::EObject* AbstractProxyResolverService::getEObject(const emf::common::URI& /*uri*/, emf::ecore::EClass* /*cls*/,
                                                                emf::sphinx::resource::ExtendedResourceSet* /*rs*/, void* /*ctx*/,
                                                                bool /*loadOnDemand*/) {
    return nullptr;
}

}  // namespace emf::sphinx::ecore::proxymanagement
